### DISCUSSION TITLE
[Bug]: absproxy strips /absproxy/<port> from request path

### DISCUSSION THREAD

**Discussion by sergey-koryshev** [coder/code-server#5439] [Q&A]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS/Web Information

- Web Browser: Chrome
- Local OS: Windows 11
- Remote OS: Raspbian GNU/Linux 11 (bullseye) 
- Remote Architecture: ARM
- `code-server --version`: 4.5.1

### Steps to Reproduce

1. open code-server
2. create an empty angular application (no routing, SCSS)
3. start it with command `ng serve --disable-host-check`
4. open link `<code-server-path>/absproxy/4200`

### Expected

Application is correctly opened in browser

### Actual

Application is not opened in browser properly because of `absproxy` still strips `/absproxy/<port>` from request path (but [docs](https://coder.com/docs/code-server/latest/guide#stripping-proxyport-from-the-request-path) says that `absproxy` will keep `/absproxy/<port>`)

### Screenshot/Video

![image](https://user-images.githubusercontent.com/11807074/183021085-cb19712a-abb6-452a-b37f-49c670b87b78.png)

### Does this issue happen in VS Code or GitHub Codespaces?

- [X] I cannot reproduce this in VS Code.
- [X] I cannot reproduce this in GitHub Codespaces.

### Are you accessing code-server over HTTPS?

- [X] I am using HTTPS.

**Comment by jsjoeio** [comment]
> absproxy still strips /absproxy/<port> from request path

I believe this behavior is expected (cc @code-asher to confirm).

I am guessing you need to configure a proxy similar to how people do it for React/Vue
- https://coder.com/docs/code-server/latest/guide#proxying-to-create-a-react-app
- https://coder.com/docs/code-server/latest/guide#proxying-to-a-vue-app
- https://angular.io/guide/build#proxying-to-a-backend-server

**Reply by code-asher** [reply]
Hmm `/absproxy` is not supposed to strip I believe (while `/proxy` does).

**Comment by sergey-koryshev** [comment]
> > absproxy still strips /absproxy/ from request path
> 
> I believe this behavior is expected (cc @code-asher to confirm).
> 
> I am guessing you need to configure a proxy similar to how people do it for React/Vue
> 
> * https://coder.com/docs/code-server/latest/guide#proxying-to-create-a-react-app
> * https://coder.com/docs/code-server/latest/guide#proxying-to-a-vue-app
> * https://angular.io/guide/build#proxying-to-a-backend-server

I will appreciate it if you can help me to figure out how to set up correct proxy for angular app

**Comment by jsjoeio** [comment]
Sadly I've never used Angular so I wouldn't be much help. I'd read the Angular guide and give that a try.

Maybe something like: 

```js
  devServer: {
    proxy: {
      '/': 'http://localhost:8080/absproxy/4200',
    },
  },
```

**Comment by sergey-koryshev** [comment]
I tried to use this proxy configuration but its behavior is odd:

```json
{
  "/": {
    "target": "http://localhost:8080/absproxy/4200",
    "logLevel": "debug"
  }
}
```

Logs:
```
[HPM] GET /absproxy/4200 -> http://localhost:8080/absproxy/4200
[HPM] GET /absproxy/4200/absproxy/4200 -> http://localhost:8080/absproxy/4200
[HPM] GET /absproxy/4200/absproxy/4200/absproxy/4200 -> http://localhost:8080/absproxy/4200
[HPM] GET /absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200 -> http://localhost:8080/absproxy/4200
[HPM] GET /absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200 -> http://localhost:8080/absproxy/4200
[HPM] GET /absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200 -> http://localhost:8080/absproxy/4200
...
[HPM] GET /absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200/absproxy/4200 -> http://localhost:8080/absproxy/4200
```

Due to a lot of the redirections my request fails with error `ERR_TOO_MANY_REDIRECTS` in browser

**Reply by jsjoeio** [reply]
Hmm...if you provide repro steps or a repo to try locally, I can try (can you also include steps from 0? I'm not familiar with Angular so assume I know nothing)

**Reply by code-asher** [reply]
There might be some kind of base path/url that can be set?

https://angular.io/guide/deployment#the-deploy-url

**Reply by sergey-koryshev** [reply]
I created a repo you can try locally: https://github.com/sergey-koryshev/angular-proxy-issue. Use command `npm run start-code-server`to start local angular dev-server.

**Reply by jsjoeio** [reply]
## Notes

### using localhost:4200 works 

I'm running code-server 4.5.1 locally on macOS + Brave and it works obviously when i access directly at localhost:4200

<img width="912" alt="image" src="https://user-images.githubusercontent.com/3806031/184027035-59881feb-dce3-4206-a414-e1eef8d66b11.png">

### using /proxy/4200 it fails

<img width="912" alt="image" src="https://user-images.githubusercontent.com/3806031/184027168-b98ebad5-e026-453f-b5c8-71c32bdf49ad.png">

### using /absproxy/4200 fails

the requests are being made against the root instead of relative

<img width="912" alt="image" src="https://user-images.githubusercontent.com/3806031/184027256-3bbf6b35-effe-43a9-a111-4b38d8cbc175.png">

Tried using `  providers: [{ provide: APP_BASE_HREF, useValue: '/absproxy/4200' }],` per [docs](https://angular.io/api/common/APP_BASE_HREF) but no luck. 

Also tried using ` <base href="./">` in `src/index.html`. The request is correct but it returns a 404:

<img width="1712" alt="image" src="https://user-images.githubusercontent.com/3806031/184030217-18a03815-1b60-429b-b10f-e22ac8302d0e.png">

Also tried using `--public-host http://localhost:8080/absproxy/4200` in the `start-code-server` script but no luck either  🤔 Hmm

Resources:
- https://angular.io/cli/serve#options
- https://github.com/angular/angular-cli/issues/19196
- https://stackoverflow.com/questions/48347430/dynamic-location-of-assets-in-angular

**Reply by jsjoeio** [reply]
Okay I got it working! [PR with changes](https://github.com/sergey-koryshev/angular-proxy-issue/pull/1).

<img width="1712" alt="image" src="https://user-images.githubusercontent.com/3806031/184032284-b9cb5d0a-04ce-455d-b366-fd9f5b73887b.png">

Basically two changes:
- use `<base href="./">` in `src/index.html`
- add `--serve-path /absproxy/4200` to the `ng serve` command

**Reply by sergey-koryshev** [reply]
Magic! It works perfectly. The only thing I'd mention: trailing slash in `https://localhost:8080/absproxy/4200/` is required to get it work.

I think your fix is need to be placed in docs somewhere under [Accessing web services](https://coder.com/docs/code-server/latest/guide#accessing-web-services)

Thank you for your help 🙂

**Reply by jsjoeio** [reply]
Yes! I don't know why the trailing slash is required 🤔 @code-asher might. But yes, feel free to add to the docs or I can next week! 

Happy we figured it out :)

**Reply by code-asher** [reply]
If the application has relative URLs anywhere the trailing slash would make a difference but otherwise either one should work in theory.

**Reply by PaulAtkins88** [reply]
Would you guys have advise on how to get this to work with a nx generated nextjs app? <code-server-host.com>/absproxy/4200 just returns 404 not found :(

**Reply by code-asher** [reply]
I would check the Next.js docs for how to set a base path, and set it to `/absproxy/4200`.

**Reply by bilogic** [reply]
> Magic! It works perfectly. The only thing I'd mention: trailing slash in `https://localhost:8080/absproxy/4200/` is required to get it work.
> 
> I think your fix is need to be placed in docs somewhere under [Accessing web services](https://coder.com/docs/code-server/latest/guide#accessing-web-services)
> 
> Thank you for your help 🙂

When accessing your app this way, are you able to get a step debugger in VSCode?
